box-share-sdk/src/main/java/com/box/androidsdk/share/vm/InviteCollaboratorsVM.java

```diff
@@ -0,0 +1,274 @@
+            int errorStrCode = R.string.box_sharesdk_generic_error;
+            if (responseCode == HttpURLConnection.HTTP_FORBIDDEN) {
+                errorStrCode = R.string.box_sharesdk_insufficient_permissions;
+            } else if (boxException.getErrorType() == BoxException.ErrorType.NETWORK_ERROR) {
+                errorStrCode = R.string.box_sharesdk_network_error;
+            }
+            data.failure(errorStrCode);
+        }
+        return data;
+    }
+    /**
+     * Helper method for transforming BoxResponse to UI Model for addCollabsApi
+     * @param response the response to transform
+     * @return the transformed model
+     */
+    private static InviteCollaboratorsDataWrapper createAddCollabsItemData(BoxResponse<BoxResponseBatch> response) {
+        return handleCollaboratorsInvited(response.getResult());
+    }
+
+    /**
+     * Helper method for helping with transformation of BoxBatchResponse to UI Model
+     * @param responses the batch response to transform
+     * @return the transformed model
+     */
+    private static InviteCollaboratorsDataWrapper handleCollaboratorsInvited(BoxResponseBatch responses) {
+        int strCode = R.string.box_sharesdk_generic_error; //default generic error
+        boolean mInvitationFailed;
+        String subMssg;
+
+        int alreadyAddedCount = 0;
+        boolean didRequestFail = false;
+        String name = "";
+        HashSet<Integer> failureCodes = generateFailureCodes();
+        List<String> failedCollaboratorsList = new ArrayList<>();
+
+
+        for (BoxResponse<BoxCollaboration> r : responses.getResponses()) {
+            if (!r.isSuccess()) {
+                if (checkIfKnownFailure(r, failureCodes)) {
+                    name = updateFailureStats(r, failedCollaboratorsList);
+                    alreadyAddedCount += name != null ? 1 : 0;
+                }
+                didRequestFail = true;
+            }
+        }
+
+        if (didRequestFail) {
+            String[] result = processAddCollabsRequestFailure(failedCollaboratorsList, name, alreadyAddedCount);
+            strCode = Integer.parseInt(result[0]);
+            subMssg = result[1];
+        } else {
+            String[] result = processAddCollabsRequestSuccess(responses);
+            strCode = Integer.parseInt(result[0]);
+            subMssg = result[1];
+        }
+
+        mInvitationFailed = (didRequestFail && !failedCollaboratorsList.isEmpty());
+
+        return new InviteCollaboratorsDataWrapper(subMssg, strCode, mInvitationFailed);
+    }
+
+    /**
+     * A helper method for updating attributes used for keeping track of stats for failure
+     * @param r the response to get request infos from
+     * @param failedCollaboratorsList list of failed collabs to be updated
+     * @return the name of collaborator that is already added
+     */
+    @VisibleForTesting
+    static String updateFailureStats(BoxResponse<BoxCollaboration> r, List<String> failedCollaboratorsList) {
```

Comment:
This is a strange method - I'd put failedCollaborators and repeatedCollaborators in a single structure that is returned by this method; and have a method getFailureDetails(BoxResponse<BoxCollaboration> r)

Reply:
I agree. I will probably redo this whole method. I was trying to just copy and paste from legacy code and putting it into a different method.

Reply:
But I have unit tests now. So I will know if I break something!